src/fix/main.c

```diff
@@ -65,535 +75,944 @@ static void print_usage(void)
+			goto free_romx;
+		} else if (writeLen < totalRomxLen) {
+			report("FATAL: Could only write %ld of \"%s\"'s %ld ROMX bytes\n",
+			       writeLen, name, totalRomxLen);
+			goto free_romx;
+		}
+	}
 
-		header[0x4B] = licensee;
+	// Output padding
+	if (padValue != 0x100) {
+		if (input == output) {
+			if (lseek(output, 0, SEEK_END) == (off_t)-1) {
+				report("FATAL: Failed to seek to end of \"%s\": %s\n",
+				       name, strerror(errno));
+				goto free_romx;
+			}
+		}
+		memset(bank, padValue, sizeof(bank));
+		size_t len = (nbBanks - 1) * BANK_SIZE - totalRomxLen; // Don't count ROM0!
+
+		while (len) {
+			static_assert(sizeof(bank) <= SSIZE_MAX);
+			size_t thisLen = len > sizeof(bank) ? sizeof(bank) : len;
+			ssize_t ret = writeBytes(output, bank, thisLen);
+
+			if (ret != thisLen) {
+				report("FATAL: Failed to write \"%s\"'s padding: %s\n",
+				       name, strerror(errno));
+				break;
+			}
+			len -= thisLen;
+		}
 	}
 
-	if (setversion) {
-		/*
-		 * Offset 0x14C: Mask ROM Version Number
-		 * Which version of the ROM this is.
-		 */
+free_romx:
+	free(romx);
+}
+
+#undef trySeek
 
-		header[0x4C] = version;
+static bool processFilename(char const *name)
+{
+	nbErrors = 0;
+	if (!strcmp(name, "-")) {
+		name = "<stdin>";
+		processFile(STDIN_FILENO, STDOUT_FILENO, name, 0);
+
+	} else {
+		// POSIX specifies that the results of O_RDWR on a FIFO are undefined.
+		// However, this is necessary to avoid a TOCTTOU, if the file was changed between
+		// `stat()` and `open(O_RDWR)`, which could trigger the UB anyway.
+		// Thus, we're going to hope that either the `open` fails, or it succeeds but IO
+		// operations may fail, all of which we handle.
+		int input = open(name, O_RDWR);
+		struct stat stat;
+
+		if (input == -1) {
+			fprintf(stderr, "FATAL: Failed to open \"%s\" for reading+writing: %s\n",
+				name, strerror(errno));
+			nbErrors = 1;
+			goto fail;
+		}
+		if (fstat(input, &stat) == -1) {
+			close(input);
+			fprintf(stderr, "FATAL: Failed to stat \"%s\": %s\n",
+				name, strerror(errno));
+			nbErrors = 1;
+			goto fail;
+		}
+		// FIXME: Do we want to support other file types?
+		if (!S_ISREG(stat.st_mode)) {
+			close(input);
+			fprintf(stderr,
+				"FATAL: \"%s\" is not a regular file, and thus cannot be modified in-place\n",
+				name);
+			nbErrors = 1;
+			goto fail;
+		}
+
+		processFile(input, input, name, stat.st_size);
+
+		close(input);
 	}
+	if (nbErrors)
+fail:
+		fprintf(stderr, "Fixing \"%s\" failed with %u errors\n", name, nbErrors);
+	return nbErrors;
+}
 
-	if (fixheadsum || trashheadsum) {
-		/*
-		 * Offset 0x14D: Header Checksum
-		 */
+int main(int argc, char *argv[])
+{
+	char ch;
 
-		uint8_t headcksum = 0;
+	while ((ch = musl_getopt_long_only(argc, argv, optstring, longopts, NULL)) != -1) {
+		switch (ch) {
+			size_t len;
+#define parseByte(output, name) \
+do { \
+	char *endptr; \
+	unsigned long tmp = strtoul(optarg, &endptr, 0); \
```

Comment:
Might want to support `$00` style hex numbers next to 0x00, as that's also what is printed as hex numbers and what the rest of rgbds expects as hex format.

Reply:
Good catch!